doc/mcup.1
'\" t
.\"     Title: mcup
.\"    Author: [see the "AUTHOR(S)" section]
.\" Generator: Asciidoctor 2.0.16
.\"      Date: 2022-01-03
.\"    Manual: \ \&
.\"    Source: \ \&
.\"  Language: English
.\"
.TH "MCUP" "1" "2022-01-03" "\ \&" "\ \&"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.ss \n[.ss] 0
.nh
.ad l
.de URL
\fI\\$2\fP <\\$1>\\$3
..
.als MTO URL
.if \n[.g] \{\
.  mso www.tmac
.  am URL
.    ad l
.  .
.  am MTO
.    ad l
.  .
.  LINKSTYLE blue R < >
.\}
.SH "NAME"
mcup \- Command line tool to keep your local maven repository small and tidy.
.SH "SYNOPSIS"
.sp
\fBmcup\fP [\fIFLAGS\fP] [\fIOPTIONS\fP] \fISUBCOMMAND\fP
.SH "DESCRIPTION"
.sp
Maven is a great tool to build software for Java and other languages running on the JVM. At its core Maven is a dependency manager which downloads and stores dependencies in a local repository. Over time, this repository grows and takes up more and more space on the hard disk. It often contains obsolete versions or dependencies that are no longer needed.
.sp
The mcup(1) command helps you to clean up your local repository. It uses filters to select artifacts based on the Maven coordinates \fIgroupId\fP, \fIartifactId\fP and \fIversion\fP. It knows three modes:
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.  sp -1
.  IP " 1." 4.2
.\}
analyze and report the size of the artifacts selected by the filters
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.  sp -1
.  IP " 2." 4.2
.\}
remove all artifacts selected by the filters and keep the rest
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 3.\h'+01'\c
.\}
.el \{\
.  sp -1
.  IP " 3." 4.2
.\}
keep all artifacts selected by the filters and remove the rest
.RE
.SH "FLAGS"
.sp
\fB\-r, \-\-releases\fP
.RS 4
Selects released artifacts only
.RE
.sp
\fB\-s, \-\-snapshots\fP
.RS 4
Selects snapshot artifacts only
.RE
.sp
\fB\-h, \-\-help\fP
.RS 4
Prints help information
.RE
.sp
\fB\-V, \-\-version\fP
.RS 4
Prints version information
.RE
.SH "OPTIONS"
.sp
\fB\-g, \-\-groups\fP \fIGROUPS\fP
.RS 4
Selects artifacts based on the group ID. Subgroups are included by default:
.sp
.if n .RS 4
.nf
.fam C
\*(Aqorg\*(Aq                 All groups starting with \*(Aqorg\*(Aq (including \*(Aqorg\*(Aq)
\*(Aqorg.wildfly\*(Aq         All groups starting with \*(Aqorg.wildfly\*(Aq (includes \*(Aqorg.wildfly)
\*(Aqorg.wildfly.core\*(Aq    All groups starting with \*(Aqorg.wildfly.core\*(Aq (includes \*(Aqorg.wildfly.core\*(Aq)
.fam
.fi
.if n .RE
.RE
.sp
\fB\-a, \-\-artifacts\fP \fIARTIFACTS\fP
.RS 4
Selects artifacts based on the artifact ID. Supports globbing like in \*(Aqmaven\-*\-plugin\*(Aq.
.RE
.sp
\fB\-v, \-\-versions\fP \fIVERSIONS\fP
.RS 4
Selects artifacts based on version (ranges). Use \fI\*(Aq<n>..\*(Aq\fP to select the \fIn\fP most recent versions, \fI\*(Aq..<n>\*(Aq\fP to select the \fIn\fP oldest versions and \fI\*(Aq<version>\*(Aq\fP to select one specific version only:
.sp
.if n .RS 4
.nf
.fam C
\*(Aq2..\*(Aq    the two most recent versions
\*(Aq..3\*(Aq    the three oldest versions
\*(Aq1.2.3\*(Aq  exactly version 1.2.3
.fam
.fi
.if n .RE
.RE
.sp
\fB\-l, \-\-local\-repository\fP \fILOCAL_REPOSITORY\fP
.RS 4
Sets the location of the local maven repository. Respects the directory configured by \fI<localRepository/>\fP in \*(Aq~/.m2/settings.xml\*(Aq. Falls back to \*(Aq~/.m2/repository\*(Aq, if nothing has been specified or configured.
.sp
Thus, the location of the local repository is computed in this order:
.sp
.RS 4
.ie n \{\
\h'-04' 1.\h'+01'\c
.\}
.el \{\
.  sp -1
.  IP " 1." 4.2
.\}
The value of the option \fI\-\-local\-repository\fP
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 2.\h'+01'\c
.\}
.el \{\
.  sp -1
.  IP " 2." 4.2
.\}
The value of \fI<localRepository/>\fP in \*(Aq~/.m2/settings.xml\*(Aq
.RE
.sp
.RS 4
.ie n \{\
\h'-04' 3.\h'+01'\c
.\}
.el \{\
.  sp -1
.  IP " 3." 4.2
.\}
Fall back to \*(Aq~/.m2/repository/\*(Aq
.RE
.RE
.SH "SUBCOMMANDS"
.sp
\fBdu\fP
.RS 4
Analyzes the disk usage of the artifacts selected by the filters. The subcommand accepts the same filters as the \fIkeep\fP and \fIrm\fP subcommands, but does not remove any artifacts. Instead, it selects the artifacts matched by the filters and calculates the size of the groups, artifacts and versions.
.sp
The subcommand accepts the following options:
.sp
\fB\-o, \-\-output\fP \fIOUTPUT\fP
.RS 4
Defines whether (g)roups, (a)rtifacts and (v)ersions are included in the usage summary. Defaults to \fIga\fP.
.RE
.RE
.sp
\fBkeep\fP
.RS 4
Keeps the artifacts matched by the filters and removes the rest
.RE
.sp
\fBrm\fP
.RS 4
Removes the artifacts matched by the filters and keeps the rest
.RE
.sp
The \fIkeep\fP and \fIrm\fP subcommands accepts the following options:
.sp
\fB\-d, \-\-dry\-run\fP
.RS 4
Does not remove artifacts
.RE
.sp
\fB\-\-list\fP
.RS 4
Prints the full path to the artifacts that will be removed. Use this flag together with \fI\-\-dry\-run\fP to review or post\-process artifacts that will be removed:
.sp
.if n .RS 4
.nf
.fam C
mcup \-\-versions \*(Aq3..\*(Aq keep \-\-dry\-run \-\-list > artifacts.txt
.fam
.fi
.if n .RE
.RE
.SH "EXIT STATUS"
.sp
\fB0\fP
.RS 4
Success.
.RE
.sp
\fB1\fP
.RS 4
Unsuccessful program execution.
.RE
.sp
\fB101\fP
.RS 4
The program panicked.
.RE
.SH "FILTER COMBINATIONS"
.sp
For subcommands \fIkeep\fP and \fIrm\fP at least one of \fI\-\-releases\fP, \fI\-\-snapshots\fP, \fI\-\-groups\fP, \fI\-\-artifacts\fP or \fI\-\-versions\fP is required, where \fI\-\-releases\fP and \fI\-\-snapshots\fP are mutually exclusive.
.sp
Subcommand \fIdu\fP has the same semantics as \fIrm\fP, but doesn\(cqt require a filter.
.sp
If \fI\-\-groups\fP is specified together with any other filter, only artifacts \fIbelow\fP the matched (sub)groups are subject to the subcommands (\fIdu\fP, \fIkeep\fP or \fIrm\fP). Artifacts \fIoutside\fP the matched (sub)groups won\(cqt be touched.
.sp
The following table explains the different filter combinations and describes which artifacts are analyzed, kept or removed.
.TS
allbox tab(:);
lt lt lt lt.
T{
.sp
Filter
T}:T{
.sp
du
T}:T{
.sp
keep
T}:T{
.sp
rm
T}
T{
.sp
\fI\-\-groups\fP only
T}:T{
.sp
Analyzes the specified (sub)groups.
T}:T{
.sp
Keeps the specified (sub)groups and removes anything else.
T}:T{
.sp
Removes the specified (sub)groups.
T}
T{
.sp
\fI\-\-artifacts\fP only
T}:T{
.sp
Analyzes the specified artifacts.
T}:T{
.sp
Keeps the specified artifacts and removes anything else.
T}:T{
.sp
Removes the specified artifacts.
T}
T{
.sp
\fI\-\-versions\fP only
T}:T{
.sp
Analyzes the specified versions.
T}:T{
.sp
Keeps the specified versions and removes anything else.
T}:T{
.sp
Removes the specified versions.
T}
T{
.sp
\fI\-\-groups\fP plus any other filter
T}:T{
.sp
Analyzes the artifacts matched by the filters \fBbelow\fP the specified (sub)groups.
T}:T{
.sp
Keeps the artifacts matched by the filters \fBbelow\fP the specified (sub)groups and removes anything else.
T}:T{
.sp
Removes the artifacts matched by the filters \fBbelow\fP the specified (sub)groups and keeps anything else.
T}
T{
.sp
All other combinations w/o \fI\-\-groups\fP
T}:T{
.sp
Analyzes the artifacts matched by the filters.
T}:T{
.sp
Keeps the artifacts matched by the filters and removes anything else.
T}:T{
.sp
Removes the artifacts matched by the filters.
T}
.TE
.sp
.SH "EXAMPLES"
.sp
Get a quick overview which groups take the most space
.br
\fBmcup du \-og\fP
.sp
Show the usage of all artifacts ending with \*(Aq\-build\*(Aq. Include groups, artifacts and versions in the usage summary.
.br
\fBmcup \-\-artifacts \*(Aq*\-build\*(Aq du \-ogav\fP
.sp
Keep the three most recent versions
.br
\fBmcup \-\-versions \*(Aq3..\*(Aq keep\fP
.sp
Remove the three oldest versions
.br
\fBmcup \-\-versions \*(Aq..3\*(Aq rm\fP
.sp
Keep the latest releases (doesn\(cqt touch snapshots)
.br
\fBmcup \-\-releases \-\-version \*(Aq1..\*(Aq keep\fP
.sp
Remove all snapshots
.br
\fBmcup \-\-snapshots rm\fP
.sp
Remove all artifacts starting with group ID \*(Aqedu\*(Aq
.br
\fBmcup \-\-groups edu rm\fP
.sp
Keep the latest maven plugins. Don\(cqt remove anything outside group \*(Aqorg.apache.maven.plugins\*(Aq.
.br
\fBmcup \-\-groups \*(Aqorg.apache.maven.plugins\*(Aq \-\-versions \*(Aq1..\*(Aq keep\fP
.sp
Remove all artifacts (across all groups) starting with \*(Aqjunit\*(Aq
.br
\fBmcup \-\-artifacts \*(Aqjunit*\*(Aq rm\fP
.SH "VERSION"
.sp
0.2.3
.SH "RESOURCES"
.sp
\fBHomepage:\fP \c
.URL "https://github.com/hpehl/mcup" "" ""
.sp
\fBSource Code:\fP \c
.URL "https://github.com/hpehl/mcup" "" ""
.sp
\fBIssue Tracker:\fP \c
.URL "https://github.com/hpehl/mcup/issues/" "" ""
.SH "AUTHORS"
.sp
Harald Pehl \c
.MTO "harald.pehl\(atgmail.com" "" ""

CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

<!-- next-header -->
## [Unreleased] - ReleaseDate

This release doesn't add new features, but upgrades dependencies only:

- anyhow: 1.0.52 → 1.0.56
- clap: 3.0.0 → 3.1.6
- clap_complete: 3.0.0 → 3.1.1
- regex: 1.5.4 → 1.5.5

## [0.2.3] - 2022-01-03

This release doesn't add new features, but upgrades dependencies only:

- anyhow: 1.0.44 → 1.0.52
- clap: 3.0.0-beta.5 → 3.0.0
- clap_generate: 3.0.0-beta.5 → 3.0.0

## [0.2.2] - 2021-10-21

Upgrade to Rust 1.56.0 stable (2021 edition)

## [0.2.1] - 2021-10-21

This release doesn't add new features, but upgrades dependencies only:

- anyhow: 1.0.42 → 1.0.44
- clap: 3.0.0-beta.2 → 3.0.0-beta.5
- clap_generate: 3.0.0-beta.2 → 3.0.0-beta.5
- console: 0.14.1 → 0.15.0
- directories: 3.0.2 → 4.0.1

## [0.2.0] - 2021-08-10

### Added
Add `du` subcommand to analyze the disk usage of the selected groups, artifacts and versions.

### Changed
Move `--dry-run` and `--list` flags to `keep` and `rm` subcommands.

## [0.1.0] - 2021-08-03
First public release

<!-- next-url -->
[Unreleased]: https://github.com/hpehl/mcup/compare/v0.2.3...HEAD
[0.2.3]: https://github.com/hpehl/mcup/compare/v0.2.2...v0.2.3
[0.2.2]: https://github.com/hpehl/mcup/compare/v0.2.1...v0.2.2
[0.2.1]: https://github.com/hpehl/mcup/compare/v0.2.0...v0.2.1
[0.2.0]: https://github.com/hpehl/mcup/compare/v0.1.0...v0.2.0
[0.1.0]: https://github.com/hpehl/mcup/releases/tag/v0.1.0


completions/mcup.bash
_mcup() {
    local i cur prev opts cmds
    COMPREPLY=()
    cur="${COMP_WORDS[COMP_CWORD]}"
    prev="${COMP_WORDS[COMP_CWORD-1]}"
    cmd=""
    opts=""

    for i in ${COMP_WORDS[@]}
    do
        case "${i}" in
            "$1")
                cmd="mcup"
                ;;
            du)
                cmd+="__du"
                ;;
            help)
                cmd+="__help"
                ;;
            keep)
                cmd+="__keep"
                ;;
            rm)
                cmd+="__rm"
                ;;
            *)
                ;;
        esac
    done

    case "${cmd}" in
        mcup)
            opts="-h -V -g -a -v -l -r -s --help --version --groups --artifacts --versions --local-repository --releases --snapshots keep rm du help"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 1 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --groups)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -g)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --artifacts)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -a)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --versions)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -v)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                --local-repository)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -l)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        mcup__du)
            opts="-o -h -V --output --help --version"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                --output)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                -o)
                    COMPREPLY=($(compgen -f "${cur}"))
                    return 0
                    ;;
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        mcup__help)
            opts="<SUBCOMMAND>..."
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        mcup__keep)
            opts="-d -h -V --dry-run --list --help --version"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
        mcup__rm)
            opts="-d -h -V --dry-run --list --help --version"
            if [[ ${cur} == -* || ${COMP_CWORD} -eq 2 ]] ; then
                COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
                return 0
            fi
            case "${prev}" in
                *)
                    COMPREPLY=()
                    ;;
            esac
            COMPREPLY=( $(compgen -W "${opts}" -- "${cur}") )
            return 0
            ;;
    esac
}

complete -F _mcup -o bashdefault -o default mcup
